### PR TITLE
slepc: set --with-arpack-dir correctly

### LINKED ISSUE #20323: Installation issue: slepc
Our spack environment, based on E4S, picked up the latest version (`3.14.1`) of `slepc` with #20291.  This version did not build on ubuntu linux with `gcc@7.5.0`.  The concrete spec yaml file I attempted to install (attached) has all the specifics of the failed spec, but to summarize it was: `slepc@3.14.1 %gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 /2m5avzr`.

Here's a summary from the error log:

```
/home/software/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa/linux-ubuntu18.04-x86_64/gcc-7.5.0/python-3.8.6-szj7jukzoj34lkchuvzrzbh52kgesac7/bin/python3.8 /home/software/spack/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa/linux-ubuntu18.04-x86_64/gcc-7.5.0/petsc-3.14.1-dtldv5gbkfwz3672gw5uonhtexncgojo/share/petsc/examples/config/gmakegen.py --petsc-arch= --pkg-dir=/tmp/root/spack-stage/spack-stage-slepc-3.14.1-2m5avzreoqsbe57hcwryejaygpqssgae/spack-src --pkg-name=slepc --pkg-pkgs=sys,eps,svd,pep,nep,mfn,lme --pkg-arch=installed-arch-linux2-c-opt
gmakefile:72: recipe for target 'installed-arch-linux2-c-opt/lib/slepc/conf/files' failed
make[3]: *** [installed-arch-linux2-c-opt/lib/slepc/conf/files] Error 127
make[3]: *** Waiting for unfinished jobs....
/tmp/root/spack-stage/spack-stage-slepc-3.14.1-2m5avzreoqsbe57hcwryejaygpqssgae/spack-src/lib/slepc/conf/slepc_rules:121: recipe for target 'slepc_libs' failed
make[2]: *** [slepc_libs] Error 2
[1;31m*******************************ERROR************************************
  Error during compile, check installed-arch-linux2-c-opt/lib/slepc/conf/make.log
  Send all contents of ./installed-arch-linux2-c-opt/lib/slepc/conf to slepc-maint@upv.es
************************************************************************[0;39m[0;49m
makefile:30: recipe for target 'all' failed
make[1]: *** [all] Error 1
GNUmakefile:4: recipe for target 'all' failed
make: *** [all] Error 2
==> [2020-12-09-19:45:12.654425] Error: ProcessError: Command exited with status 2:
    'make' 'MAKE_NP=2'

4 errors found in build log:
     288    Using mpiexec: /home/software/spack/spack_path_placeholder/spack_pa
            th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_
            path_placeholder/spack_path_placeholder/spack_path_placeholder/spac
            k_path_placeholder/spack_path_placeholder/spack_path_placeholder/sp
            ack_path_placeholder/spack_path_placeholder/spack_path_placeholder/
            spack_path_placeholder/spack_path_placeholder/spack_path_placeholde
            r/spack_path_placeholder/spack_path_placeholder/spack_path_placehol
            der/spack_path_placeholder/spack_path_placeholder/spack_pa/linux-ub
            untu18.04-x86_64/gcc-7.5.0/mpich-3.3.2-j3zl3i75xss7qzyp7mgqa27ugaj6
            3g3z/bin/mpiexec
     289    ------------------------------------------
     290    Using MAKEFLAGS: -j2 -l3.0  --no-print-directory -- SLEPC_DIR=/tmp/
            root/spack-stage/spack-stage-slepc-3.14.1-2m5avzreoqsbe57hcwryejayg
            pqssgae/spack-src PETSC_DIR=/home/software/spack/spack_path_placeho
            lder/spack_path_placeholder/spack_path_placeholder/spack_path_place
            holder/spack_path_placeholder/spack_path_placeholder/spack_path_pla
            ceholder/spack_path_placeholder/spack_path_placeholder/spack_path_p
            laceholder/spack_path_placeholder/spack_path_placeholder/spack_path
            _placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa
            th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_
            path_placeholder/spack_path_placeholder/spack_path_placeholder/spac
            k_pa/linux-ubuntu18.04-x86_64/gcc-7.5.0/petsc-3.14.1-dtldv5gbkfwz36
            72gw5uonhtexncgojo PETSC_ARCH=installed-arch-linux2-c-opt MAKE_NP=2
     291    ==========================================
     292    /home/software/spack/spack_path_placeholder/spack_path_placeholder/
            spack_path_placeholder/spack_path_placeholder/spack_path_placeholde
            r/spack_path_placeholder/spack_path_placeholder/spack_path_placehol
            der/spack_path_placeholder/spack_path_placeholder/spack_path_placeh
            older/spack_path_placeholder/spack_path_placeholder/spack_path_plac
            eholder/spack_path_placeholder/spack_path_placeholder/spack_path_pl
            aceholder/spack_path_placeholder/spack_path_placeholder/spack_path_
            placeholder/spack_path_placeholder/spack_pa/linux-ubuntu18.04-x86_6
            4/gcc-7.5.0/python-3.8.6-szj7jukzoj34lkchuvzrzbh52kgesac7/bin/pytho
            n3.8 /home/software/spack/spack_path_placeholder/spack_path_placeho
            lder/spack_path_placeholder/spack_path_placeholder/spack_path_place
            holder/spack_path_placeholder/spack_path_placeholder/spack_path_pla
            ceholder/spack_path_placeholder/spack_path_placeholder/spack_path_p
            laceholder/spack_path_placeholder/spack_path_placeholder/spack_path
            _placeholder/spack_path_placeholder/spack_path_placeholder/spack_pa
            th_placeholder/spack_path_placeholder/spack_path_placeholder/spack_
            path_placeholder/spack_path_placeholder/spack_pa/linux-ubuntu18.04-
            x86_64/gcc-7.5.0/petsc-3.14.1-dtldv5gbkfwz3672gw5uonhtexncgojo/shar
            e/petsc/examples/config/gmakegen.py --petsc-arch= --pkg-dir=/tmp/ro
            ot/spack-stage/spack-stage-slepc-3.14.1-2m5avzreoqsbe57hcwryejaygpq
            ssgae/spack-src --pkg-name=slepc --pkg-pkgs=sys,eps,svd,pep,nep,mfn
            ,lme --pkg-arch=installed-arch-linux2-c-opt
     293    gmakefile:72: recipe for target 'installed-arch-linux2-c-opt/lib/sl
            epc/conf/files' failed
  >> 294    make[3]: *** [installed-arch-linux2-c-opt/lib/slepc/conf/files] Err
            or 127
     295    make[3]: *** Waiting for unfinished jobs....
     296    /tmp/root/spack-stage/spack-stage-slepc-3.14.1-2m5avzreoqsbe57hcwry
            ejaygpqssgae/spack-src/lib/slepc/conf/slepc_rules:121: recipe for t
            arget 'slepc_libs' failed
  >> 297    make[2]: *** [slepc_libs] Error 2
     298    *******************************ERROR*******************************
            *****
     299      Error during compile, check installed-arch-linux2-c-opt/lib/slepc
            /conf/make.log
     300      Send all contents of ./installed-arch-linux2-c-opt/lib/slepc/conf
             to slepc-maint@upv.es
     301    *******************************************************************
            *****
     302    makefile:30: recipe for target 'all' failed
  >> 303    make[1]: *** [all] Error 1
     304    GNUmakefile:4: recipe for target 'all' failed
  >> 305    make: *** [all] Error 2
```

### Information on your system

Output of `spack debug report`:

```
$ spack debug report
* **Spack:** 0.16.0-202-9bcf99358
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-haswell
* **Concretizer:** original
```

The build failure happened inside of the following spack environment:

[spack.yaml.txt](https://github.com/spack/spack/files/5673905/spack.yaml.txt)

The concrete `spec.yaml` I attempted to install:

[slepc.yaml.txt](https://github.com/spack/spack/files/5673981/slepc.yaml.txt)

### Additional information

Spack build logs:

[spack-build-env.txt](https://github.com/spack/spack/files/5673918/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/5673919/spack-build-out.txt)

Other logs mentioned in the build log error messages (contents of `installed-arch-linux2-c-opt/lib/slepc/conf`) are attached below.

[slepcvariables.txt](https://github.com/spack/spack/files/5673946/slepcvariables.txt)
[slepcrules.txt](https://github.com/spack/spack/files/5673947/slepcrules.txt)
[make.log.txt](https://github.com/spack/spack/files/5673948/make.log.txt)
files.txt [empty]
[error.log.txt](https://github.com/spack/spack/files/5673956/error.log.txt)
[configure.log.txt](https://github.com/spack/spack/files/5673957/configure.log.txt)
[modules/slepc/3.14.1.txt](https://github.com/spack/spack/files/5673966/3.14.1.txt)

Mentioning the maintainers of `slepc`:

@balay @joseeroman

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by eugeneswalker**
Set `--with-arpack-dir` to the `arpack.prefix` install prefix, not the lib prefix.

See https://slepc.upv.es/documentation/instal.htm for explanation of how `--with-arpack-dir` should be set.

Fixes https://github.com/spack/spack/issues/20323

### CODE REVIEW

**Comment by eugeneswalker** [comment]
This fixed my `slepc` failures on:
* `RHEL 7` `x86_64` and `ppc64le`
* `RHEL 8` `x86_64` and `ppc64le`

@scottwittenburg

**Comment by scottwittenburg** [comment]
Nice work @eugeneswalker!  Thanks!

> This fixed my `slepc` failures on:
> 
> * `RHEL 7` `x86_64` and `ppc64le`
> * `RHEL 8` `x86_64` and `ppc64le`
> * `Ubuntu 18.04` `x86_64` and `ppc64le`
> * `Ubuntu 20.04` `x86_64` and `ppc64le`

Does this mean you eventually encountered the problem on Ubuntu 18.04 and 20.04?

**Comment by eugeneswalker** [comment]
> Does this mean you eventually encountered the problem on Ubuntu 18.04 and 20.04?

Oh, actually, no. I got carried away listing those platforms. It's been a long day. I swear I'm not jerking your chain! It always built for me on Ubuntu!! I think, for whatever reason that I don't understand, arpack was using `lib64` path on RHEL 7+8, but using `lib` on Ubuntu 18.04+20.04.

**Comment by eugeneswalker** [comment]
If you approve this I'll merge it right now!

**Comment by scottwittenburg** [comment]
> If you approve this I'll merge it right now!

Let's give the maintainers a chance to look it over, but it seems ok to me, so I'm happy to approve it if we don't hear back sometime today.

**Comment by joseeroman** [comment]
I am ok with this change. The fault seems to be in SLEPc's `configure`, where I check for `lib` directories only. I will change it upstream so that it also checks for `lib64`.

I don't have access to a machine with RHEL, so I could not reproduce the problem. Thanks for fixing.

**Review by scottwittenburg — APPROVED** [review]
LGTM, thanks @eugeneswalker!

File: var/spack/repos/builtin/packages/slepc/package.py
Comment (balay) [inline var/spack/repos/builtin/packages/slepc/package.py:99]: perhaps its better to use [untested]
```
'--with-arpack-include=%s' % spec['arpack-ng'].prefix.include,
'--with-arpack-lib=%s' % spec['arpack-ng'].prefix.lib,
```

Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: well the issue i was seeing is, on rhel, spec['arpack-ng'].prefix.lib came out to ...lib even though the actual directory is lib64 on rhel. it was lib on ubuntu and lib64 on rhel.

Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: the documentation i linked to in the original PR description said that --with-arpack-dir should point to the arpack prefix, not a lib directory.

Comment (balay) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: Ok. So the primary bug is in the code that provides `arpack.prefix.lib`? so in arpack?

generally package-include/lib is preferred over package-dir option - this avoids some of the guesswork configure does [which can sometimes be wrong]

Comment (joseeroman) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: @balay The option `--with-arpack-lib=` is intended to provide a list of libraries, not a path. Isn't `arpack.prefix.lib` a path?

Comment (balay) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: Ah - sorry - should have used:
```
'--with-arpack-lib=%s' % spec['arpack-ng'].prefix.libs.joined()
```

Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: It looks like `--with-arpack-include` is not recognized. At least that's what I'm seeing when I try to use it:

```
$> spack install -f slepc
...
==> Installing slepc-3.14.1-oa2mlhnzrwmvtkr45ii7qmffkxp4t73t
==> Using cached archive: /opt/spack/var/spack/cache/_source-cache/archive/cc/cc78a15e34d26b3e6dde003d4a30064e595225f6185c1975bbd460cb5edd99c7.tar.gz
==> slepc: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.1.0/python-3.8.6-e7dxrymceexhq6cg6f3jtp6sevyozzko/bin/python3.8' 'configure' '--prefix=/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.1.0/slepc-3.14.1-oa2mlhnzrwmvtkr45ii7qmffkxp4t73t' '--with-arpack-include=/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.1.0/arpack-ng-3.8.0-hu32xolzvdqbjdtqozml4e4bgmjkwhbz/include' '--with-arpack-lib=-lparpack,-larpack'

1 error found in build log:
     1    ==> slepc: Executing phase: 'install'
     2    ==> [2020-12-18-18:23:29.750450] '/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.1.0/python-3.8.6-e7dxrymceexhq6cg6f3j
          tp6sevyozzko/bin/python3.8' 'configure' '--prefix=/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.1.0/slepc-3.14.1-oa2m
          lhnzrwmvtkr45ii7qmffkxp4t73t' '--with-arpack-include=/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.1.0/arpack-ng-3.8.
          0-hu32xolzvdqbjdtqozml4e4bgmjkwhbz/include' '--with-arpack-lib=-lparpack,-larpack'
  >> 3    Checking environment... ERROR: Invalid arguments --with-arpack-include=/opt/spack/opt/spack/linux-rhel7-x86_64/gcc-8.
          1.0/arpack-ng-3.8.0-hu32xolzvdqbjdtqozml4e4bgmjkwhbz/include
     4    Use -h for help
```

Comment (joseeroman) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: Yes, this option is not available because in SLEPc we do not use ARPACK's header files. Probably we should ignore it instead of erroring out. Try with `--with-arpack-lib=` only. Thanks.

Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: In that case, I will merge this PR then. Everything about slepc stays the same except `--with-arpack-dir` is *correctly* set to the arpack install prefix. The current package sets `--with-arpack-dir` to the lib dir of arpack, which is not right. 

I've tested this PR on every system I have my hands on, and `slepc` builds in all cases:
* RHEL 7 + 8, x86_64 and ppc64le
* Ubuntu 18.04 + 20.04, x86_64 and ppc64le (wasn't broken before, but still works with these changes)

@balay do you object to this? We can always change to `--with-arpack-include` if, in the future, that option is supported.

Comment (joseeroman) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: For reference, I have made changes in SLEPc that will be included in version 3.14.2, cf. [MR](https://gitlab.com/slepc/slepc/-/merge_requests/108).

Comment (balay) [reply var/spack/repos/builtin/packages/slepc/package.py:99]: Looks like --with-arpack-include option does not exist - so only change would be to use --with-arpack-lib. However I see some extra code in slepc/package.py - that lists the arpack libraries separately - that can likely be removed.

I'm fine with this current change for now.

**Review by balay — APPROVED** [review]

**Pull request merged**
